### PR TITLE
3915 - Cycle Switch: Public access ->Last Published only

### LINKED ISSUE #3915: Cycle Switch: Public access ->Last Published only
- [ ] Not logged users can only access and see last published cycle
- [ ] Refactor the cycle switch label so that :
  - the option label is composed by AssessmentShortName + cycle label
  - option label of the last editing cycle label is AssessmentShortName + latest (ask translations labels AssessmentShortName (FRA, panEuropean) + 'latest'
Image
  - the selected option label is the last published cycle is empty string
Image

### PR DESCRIPTION
**Pull request by sorja**
# 3915 - Cycle Switch: Public access ->Last Published only
---
This pull request will change non logged user logic:
Non logged in users will have access only to a single cycle, the latest published cycle

### Non logged user
https://github.com/user-attachments/assets/6f9b5d92-378a-4362-bb97-8bc4101e8079

### Logged admin user
https://github.com/user-attachments/assets/d42e5d45-baa8-486f-823c-e172e4d093fc

### Logged non admin user
https://github.com/user-attachments/assets/d94a31dc-45b3-408e-b9a3-4d1c4ebfa989

Developers note:
Translations added with DeepL

### CODE REVIEW

File: src/client/components/PageLayout/Header/CycleSwitcher/CycleSwitcher.tsx [hunk 1]
```diff
@@ -1,23 +1,30 @@
...
 const CycleSwitcher = () => {
+  const assessment = useAssessment()
   const { cycleName } = useCycleRouteParams()
   const popoverItems = usePopoverItems()
 
-  if (popoverItems.length < 1) return <span className="cycle-switcher-locked">{cycleName}</span>
+  if (popoverItems.length < 1) return null
+
+  const latestCycle = Assessments.getLastCreatedCycle(assessment)
+  const displayName = latestCycle.name === cycleName ? '' : cycleName
```
Comment (minotogna) [inline src/client/components/PageLayout/Header/CycleSwitcher/CycleSwitcher.tsx:21]: The empty string should be associated with the lastest editing ?

Comment (minotogna) [reply src/client/components/PageLayout/Header/CycleSwitcher/CycleSwitcher.tsx:21]: Nope sorry . Latest created

File: src/client/components/PageLayout/Header/CycleSwitcher/hooks/usePopoverItems.tsx [hunk 2]
```diff
@@ -33,11 +33,14 @@ export const usePopoverItems = (): Array<PopoverItem> => {
 
           const assessmentName = assessment.props.name
           const cycleName = cycle.name
+          const isLatestCycle = Assessments.getLastCreatedCycle(assessment)?.name === cycleName
           const isCurrentRoute = assessmentName === routeParams.assessmentName && cycleName === routeParams.cycleName
+          const label = t(`${assessmentName}.labels.short`)
+          const content = `${label} ${isLatestCycle ? t('common.latest') : cycleName}`
```
Comment (minotogna) [inline src/client/components/PageLayout/Header/CycleSwitcher/hooks/usePopoverItems.tsx:39]: Please check pan eu works as well

File: src/meta/assessment/assessments.ts [hunk 3]
```diff
@@ -55,6 +56,49 @@ export const AssessmentStatusTransitions = {
...
 const getShortLabel = (assessmentName: AssessmentName) => `${assessmentName}.labels.short`
 
+/**
+ * Retrieves the most recently published cycle from an assessment.
+ *
+ * @param {Assessment} assessment - Assessment
+ * @returns {Cycle | undefined} The most recently published cycle, or undefined if no published cycles exist.
+ */
+const getLastPublishedCycle = (assessment: Assessment): Cycle | undefined => {
+  const publishedCycles = assessment.cycles.filter((cycle) => cycle.props.status === CycleStatus.published)
```
Comment (minotogna) [inline src/meta/assessment/assessments.ts]: I would use Cycles.isPublished

File: src/meta/assessment/assessments.ts [hunk 4]
```diff
@@ -55,6 +56,49 @@ export const AssessmentStatusTransitions = {
...
+
+  if (publishedCycles.length === 0) {
+    return undefined
+  }
+
+  return publishedCycles.reduce((last, current) => {
+    const lastDate = new Date(last.props.datePublished)
+    const currentDate = new Date(current.props.datePublished)
+    return lastDate > currentDate ? last : current
```
Comment (minotogna) [inline src/meta/assessment/assessments.ts]: Do we have Dates utility to compare dates ?

File: src/meta/assessment/assessments.ts [hunk 5]
```diff
@@ -55,6 +56,49 @@ export const AssessmentStatusTransitions = {
...
+
+/**
+ * Retrieves the most recently created cycle from an assessment.
+ * Note: cycle.props.status can be draft, even when dateCreated is present.
+ *
+ * @param {Assessment} assessment - Assessment
+ * @returns {Cycle | undefined} The most recently created cycle, or undefined if no cycles exist.
+ */
+const getLastCreatedCycle = (assessment: Assessment): Cycle | undefined => {
+  const createdCycles = assessment.cycles.filter((cycle) => cycle.props.dateCreated)
```
Comment (minotogna) [inline src/meta/assessment/assessments.ts]: Can there ever be a cycle that hasn't the created date ?

File: src/meta/assessment/assessments.ts [hunk 6]
```diff
@@ -55,6 +56,49 @@ export const AssessmentStatusTransitions = {
...
+
+  if (createdCycles.length === 0) {
+    return undefined
+  }
+
+  return createdCycles.reduce((last, current) => {
+    const lastDate = new Date(last.props.dateCreated)
+    const currentDate = new Date(current.props.dateCreated)
+    return lastDate > currentDate ? last : current
```
Comment (minotogna) [inline src/meta/assessment/assessments.ts]: Same as above for comparing dates

File: src/meta/assessment/assessments.ts (hunk 5 above)
Comment (sorja) [reply src/meta/assessment/assessments.ts]: That's the thing I was wondering also, at least not now in our current setup
I will remove it anyway, let me know if there can be (eg 2030 draft, but not created)

File: src/client/components/PageLayout/Header/CycleSwitcher/hooks/usePopoverItems.tsx (hunk 2 above)
Comment (sorja) [reply src/client/components/PageLayout/Header/CycleSwitcher/hooks/usePopoverItems.tsx:39]: Tested in the videos, let me know if I should check some other case I missed

image

File: src/meta/assessment/assessments.ts (hunk 5 above)
Comment (minotogna) [reply src/meta/assessment/assessments.ts]: No , I think a cycle to exist must be created first . Yes , let's just remove this check for now 😉

**Review by minotogna — APPROVED** [review]
👍

**Pull request merged**